src/fairseq2/datasets/speech.py

```diff
@@ -0,0 +1,332 @@
+        max_audio_len: int,
+        max_num_elements: int,
+        *,
+        dtype: DataType = torch.float32,
+        min_audio_len: int = 1,
+        normalize_audio: bool = False,
+        example_shuffle_window: int = 1,
+        batch_shuffle_window: int = 1,
+        num_accumulate: int = 1,
+        num_prefetch: int = 1,
+        seed: int = 2,
+        **extras: Any,
+    ) -> DataReader[Tensor]:
+        """Create a dataset reader.
+
+        :param split:
+            The split to read.
+        :param gang:
+            The gang over which to shard the dataset.
+        :param max_audio_len:
+            The maximum audio length of each example. Examples longer than
+            this value will be cropped.
+        :param max_num_elements:
+            The maximum number of elements in each batch.
+        :param dtype:
+            The data type of the decoded audio sequences.
+        :param min_audio_len:
+            The minimum audio length of each example. Examples shorter than
+            this value will be dropped.
+        :param normalize_audio:
+            If ``True``, normalizes audio to have zero mean and unit variance.
+        :param example_shuffle_window:
+            The size of the sliding window for shuffling examples. If ``1``, no
+            shuffling is performed; if ``0``, true shuffling is performed by
+            loading the entire dataset.
+        :param batch_shuffle_window:
+            The size of the sliding window for shuffling batches. If ``1``, no
+            shuffling is performed; if ``0``, true shuffling is performed by
+            loading the entire dataset.
+        :param num_accumulate:
+            The number of batches to accumulate in each iteration. Typically
+            used with gradient accumulation during training.
+        :param num_prefetch:
+            The number of batches to prefetch in background.
+        :param seed:
+            The seed to initialize the random number generators used internally.
+        :param extras:
+            The extra parameters specific to the dataset implementation.
+        """
+
+    @abstractmethod
+    def splits(self) -> Set[str]:
+        """Return the set of splits."""
+
+
+load_speech_dataset = DelegatingDatasetLoader[SpeechDataset]()
+
+
+# TODO: FIX, INFER
+npc = 10
+
+
+# TODO: Work in progress!
+@final
+class GenericSpeechDataset(SpeechDataset):
+    """Represents a generic manifest-based Speech dataset."""
+
+    _dataset_name: str
+    _manifest_dir: Path
+    _splits: Set[str]
+
+    def __init__(self, dataset_name: str, manifest_dir: Path) -> None:
+        """
+        :param dataset_name:
+            The name of the dataset.
+        :param manifest_dir:
+            The directory under which the manifest files resides.
+        """
+        self._dataset_name = dataset_name
+        self._manifest_dir = manifest_dir
+
+        self._splits = set()
+
+        for tsv_file in manifest_dir.glob("*.tsv"):
```

Comment:
This thing force to use all tsv in the folder, it there an option to load only few of them, either for debugging or for specific runs